Base/BusyThreads.hpp

```diff
@@ -32,18 +37,33 @@
 class BusyThreads
 {
 public:
+  using Seconds = std::chrono::duration<double>;
+
   BusyThreads();
   ~BusyThreads();
 
   int numThreads() const;
   void setNumThreads(int numThreads);
 
+  //! The duration of one busy thread iteration.
+  Seconds period() const;
+  void setPeriod(Seconds period);
+
+  //! The percentage of a busy thread iteration spent performing low-energy work rather
+  //! than blocking.
+  double threadCpuUsage() const;
+  void setThreadCpuUsage(double threadCpuUsage);
+
 private:
   void setup(int numThreads);
   void teardown();
 
   void busyThread(int threadIndex);
 
-  std::atomic<bool> mIsActive{false};
   std::vector<std::thread> mThreads;
+  std::condition_variable mConditionVariable;
+  std::mutex mMutex;
+  std::atomic<bool> mIsActive{false};
```

Comment:
Probably doesn't really matter in this case, but the packing here isn't great.

More significantly though, I'm wondering if the atomic `mIsActive` is really necessary.  It's only accessed when guarded by the mutex, except in the condition predicate.  It might be easier to just not use a predicate there, and write the spurious wakeup loop manually?  Then, its accesses (and essentially everything) would be guarded by the mutex and there wouldn't be a need for a mutex and also atomics.

Reply:
It can take up to 100ms to join threads if `mIsActive` isn't checked in the hardware delay loop. I don't want to use a unique lock in that loop since it would prevent other busy threads from making progress. C++17 has `std::shared_lock` now, but an atomic seems like the faster option here.

Reply:
Isn't the condition explicitly signaled in that case though?

Reply:
Yes, but if the CPU usage is set to 100%, we would block in the hardware delay loop without checking `mIsActive`.

Reply:
Oh, you mean in the `while` condition, I was just looking at the condition variable part.

I don't know, it seems a little smelly to be both constantly checking a condition variable and constantly checking an atomic, but I guess it doesn't really matter for a thread that is constantly giving up its time slice and burning time on purpose anyway.